### PR TITLE
avm2: added flash.util.escapeMultiByte

### PR DESCRIPTION
**Pull request by dowgird**
Two functions from flash.util, which are needed by flash.net.URLVariables

### CODE REVIEW

**Comment by Herschel** [comment]
I suspect that Flash isn't too strict with its URL encoding and we'll have to do this ourselves instead of relying on `urlencoding` crate.

For example,
`trace(flash.utils.escapeMultiByte("@*_+-./"));`
traces `%40%2A%5F%2B%2D%2E%2F` in Flash, but `%40%2A_%2B-.%2F` with this PR.

We should also check to see how this differs from [AVM1 `escape`](https://github.com/ruffle-rs/ruffle/blob/b18329c8feb06d4b2feaba64687abbf580599952/core/src/avm1/globals.rs#L376) -- possibly we can share code, although it wouldn't surprise me if AVM2 behavior is different here. Should probably add a TODO noting about `System.useCodePage` as well.

**Comment by EmperorBale** [comment]
> We should also check to see how this differs from [AVM1 `escape`](https://github.com/ruffle-rs/ruffle/blob/b18329c8feb06d4b2feaba64687abbf580599952/core/src/avm1/globals.rs#L376) -- possibly we can share code, although it wouldn't surprise me if AVM2 behavior is different here.

I think it's more likely the same behavior as [AVM2 `escape`](https://github.com/ruffle-rs/ruffle/blob/master/core/src/avm2/globals/toplevel.rs#L92-L123)

**Comment by dowgird** [comment]
> > We should also check to see how this differs from [AVM1 `escape`](https://github.com/ruffle-rs/ruffle/blob/b18329c8feb06d4b2feaba64687abbf580599952/core/src/avm1/globals.rs#L376) -- possibly we can share code, although it wouldn't surprise me if AVM2 behavior is different here.
> 
> I think it's more likely the same behavior as [AVM2 `escape`](https://github.com/ruffle-rs/ruffle/blob/master/core/src/avm2/globals/toplevel.rs#utf8_percent_encode("foo bar?", NON_ALPHANUMERIC).to_string()L92-L123)

So I found out that these differ. Toplevel escape skips '.' and '/' and also has different output on the "alien" multibyte char. I haven't tested avm1 escape but from looking at the code it doesn't support multibyte either. 

I'm trying to get the desired behavior from `percent_encoding` module.

**Comment by EmperorBale** [comment]
> I'm trying to get the desired behavior from `percent_encoding` module.

I'm not sure if using a crate for this is a good idea for 2 reasons:
1. We use a custom string format (WString) which is different from what a 3rd party crate would be expecting (most likely a string). That means we would have to convert WString to String, which can only be done lossy. Also keep in mind that WString is UTF16.
2. In the future were going to have to customize the character encoding output, which is not something that I think can be done with the `percent_encoding` crate, or any other crate for this.

I think a custom percent encoder would be much better.

**Comment by dowgird** [comment]
Fair points, I'll go with a custom encoder then.

> 1. We use a custom string format (WString) which is different from what a 3rd party crate would be expecting (most likely a string). That means we would have to convert WString to String, which can only be done lossy. Also keep in mind that WString is UTF16.

I presume the loss conversion occurs when WString is invalid utf-8. For triggering this in tests, would ByteArray with the invalid utf-8 work after `coerce_to_string` is called? 

I see that some tests do use ByteArray, I'm just not 100% sure if the context is the same.

**Comment by EmperorBale** [comment]
> I presume the loss conversion occurs when WString is invalid utf-8. For triggering this in tests, would ByteArray with the invalid utf-8 work after `coerce_to_string` is called?

It should work fine, you can probably take some of the code from [this test](https://github.com/ruffle-rs/ruffle/blob/master/tests/tests/swfs/avm2/invalid_utf8/Test.as), and replace `escape` with `escapeMultiByte`.

**Comment by dowgird** [comment]
Actually, `escapeMultiByte` is lossy (i.e. unpaired surrogates become the unknown char). This means we could go with `String` but I went with manually encoding anyway, as this is just simpler.

I also removed `unescapeMultiByte` from this PR, as I haven't yet fully figured out the edge cases.

File: core/src/avm2/globals/flash/utils.rs [hunk 1]
```diff
@@ -111,6 +114,32 @@ pub fn clear_timeout<'gc>(
...
+    let bs = s.as_wstr().to_utf8_lossy();
+    let mut buf = WString::new();
+    for b in bs.as_bytes() {
+        if *b == 0 {
+            break;
+        }
+        if b.is_ascii_alphanumeric() {
+            buf.push_char(*b as char);
+        } else {
+            write!(&mut buf, "%{b:X}")?;
```
Comment (Herschel) [inline core/src/avm2/globals/flash/utils.rs:136]: This needs a leading zero for <16:
```
trace(flash.utils.escapeMultiByte("\u0001")); // %01 in Flash, %1 in this PR
```
I think `{b:02X}` does the trick.

Comment (dowgird) [reply core/src/avm2/globals/flash/utils.rs]: Indeed! Fixed and added a test which would have caught this.

**Comment by Herschel** [comment]
Thank you!

**Pull request merged**